### PR TITLE
Compression implementation

### PR DESCRIPTION
**Pull request by dgquintas**
Including the compress filter for outgoing data and the changes to surface/call.c, frame_data.c and stream_encoder.c for incoming data.

### CODE REVIEW

**Comment by dgquintas** [comment]
I want to add some more tests, specifically exercising the metadata-driven compression hinting (a request goes out requesting that the response be compressed with a certain algorithm).

File: src/core/channel/compress_filter.c [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+        break;
+      case GRPC_OP_METADATA:
+        grpc_metadata_batch_add_head(
+            &(sop->data.metadata), &calld->compression_algorithm_storage,
+            grpc_mdelem_ref(channeld->mdelem_compression_algorithms
+                                [did_compress ? calld->compression_algorithm
+                                              : GRPC_COMPRESS_NONE]));
+        break;
+      case GRPC_OP_SLICE:
+        if (did_compress) {
```
Comment (ctiller) [inline src/core/channel/compress_filter.c]: What if compression generates more (or less) slices than there were originally?

File: src/core/transport/chttp2/stream_encoder.c [hunk 2]
```diff
@@ -497,7 +497,7 @@ gpr_uint32 grpc_chttp2_preencode(grpc_stream_op *inops, size_t *inops_count,
            through - this lets us reuse the slice framing code below */
         slice = gpr_slice_malloc(5);
         p = GPR_SLICE_START_PTR(slice);
-        p[0] = 0;
+        p[0] = !!(op->data.begin_message.flags & GRPC_WRITE_INTERNAL_COMPRESS);
```
Comment (ctiller) [inline src/core/transport/chttp2/stream_encoder.c]: lets call out 0, 1 explicitly here (as named constants)

Comment (dgquintas) [inline src/core/transport/chttp2/stream_encoder.c]: Done.

File: src/core/channel/compress_filter.c (hunk 1 above)
Comment (dgquintas) [inline src/core/channel/compress_filter.c]: Refactored and addressed (see new comments).

The "if" is also incidentally outside the loop, which may result in better branching behavior (but that wasn't the point of the refactoring, rather to make the code more readable).

**Comment by dgquintas** [comment]
I'm aware the PR isn't in a mergeable state. I'll merge with master once things LG.

**Comment by dgquintas** [comment]
Ping :)  There's another PR (for the accept-encoding) that builds on top of this ready for prime PR time.

**Comment by ctiller** [comment]
Needs a merge with HEAD

**Comment by dgquintas** [comment]
Merged.

File: src/core/channel/compress_filter.c [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+#include <grpc/compression.h>
+#include <grpc/support/log.h>
+#include <grpc/support/slice_buffer.h>
+
+#include "src/core/channel/compress_filter.h"
+#include "src/core/channel/channel_args.h"
+#include "src/core/compression/message_compress.h"
+
+typedef struct call_data {
+  gpr_slice_buffer slices;
```
Comment (ctiller) [inline src/core/channel/compress_filter.c]: Can you add some doc strings for the fields in this struct and in channel_data

**Comment by ctiller** [comment]
One comment request, otherwise LGTM

**Pull request merged**